### PR TITLE
Fix/crashes reported in 345 344

### LINKED ISSUE #344: "Fatal Python error: Aborted" in test_platform_creation
Test:
dpctl/tests/test_sycl_platform.py::test_platform_creation[opencl:gpu-check_name]

Environment:
```bash
runner@fv-az137-563:~/work/dpctl/dpctl$ clinfo -l
Platform #0: Intel(R) OpenCL
 `-- Device #0: Intel(R) Xeon(R) CPU E5-2673 v3 @ 2.40GHz
```

gdb:
```gdb
Thread 1 "python" received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50      ../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff7a81859 in __GI_abort () at abort.c:79
#2  0x00007ffff3a3d951 in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00007ffff3a4947c in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007ffff3a494e7 in std::terminate() () from /lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007ffff3a49799 in __cxa_throw () from /lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00007ffff5b6a26f in cl::sycl::ONEAPI::detail::filter_selector_impl::operator()(cl::sycl::device const&) const () from /opt/intel/oneapi/compiler/2021.1.1/linux/lib/libsycl.so.5
#7  0x00007ffff5c01b7a in cl::sycl::device_selector::select_device() const () from /opt/intel/oneapi/compiler/2021.1.1/linux/lib/libsycl.so.5
#8  0x00007ffff5c02f30 in cl::sycl::ONEAPI::filter_selector::select_device() const () from /opt/intel/oneapi/compiler/2021.1.1/linux/lib/libsycl.so.5
#9  0x00007ffff5c1104d in cl::sycl::platform::platform(cl::sycl::device_selector const&) () from /opt/intel/oneapi/compiler/2021.1.1/linux/lib/libsycl.so.5
#10 0x00007ffff5c7cdb7 in DPCTLPlatform_CreateFromSelector () from /home/runner/work/dpctl/dpctl/dpctl/libDPCTLSyclInterface.so
#11 0x00007ffff3055505 in __pyx_f_5dpctl_14_sycl_platform_12SyclPlatform__init_from_selector (__pyx_v_self=0x7fffdfbf54f0, __pyx_v_DSRef=<optimized out>) at dpctl/_sycl_platform.cpp:2000
#12 0x00007ffff3055207 in __pyx_f_5dpctl_14_sycl_platform_12SyclPlatform__init_from_cstring (__pyx_v_self=0x7fffdfbf54f0, __pyx_v_string=<optimized out>) at dpctl/_sycl_platform.cpp:1943
#13 0x00007ffff30588a0 in __pyx_pf_5dpctl_14_sycl_platform_12SyclPlatform_2__cinit__ (__pyx_v_arg=<optimized out>, __pyx_v_self=0x7fffdfbf54f0) at dpctl/_sycl_platform.cpp:2453
#14 __pyx_pw_5dpctl_14_sycl_platform_12SyclPlatform_3__cinit__ (__pyx_kwds=<optimized out>, __pyx_args=<optimized out>, __pyx_v_self=0x7fffdfbf54f0) at dpctl/_sycl_platform.cpp:2379
#15 __pyx_tp_new_5dpctl_14_sycl_platform_SyclPlatform (t=<optimized out>, a=<optimized out>, k=<optimized out>) at dpctl/_sycl_platform.cpp:3854
#16 0x00007ffff7d91a8a in type_call (type=type@entry=0x7ffff305e280 <__pyx_type_5dpctl_14_sycl_platform_SyclPlatform>, args=args@entry=0x7fffdfbc6390, kwds=kwds@entry=0x0) at Objects/typeobject.c:951
```

### LINKED ISSUE #345: "Fatal Python error: Aborted" in test_dparray_T
Test:
dpctl/tests/test_dparray.py::Test_dparray::test_dparray_T

Environment:
```bash
runner@fv-az213-658:~/work/dpctl/dpctl$ clinfo
Number of platforms                               0
```

gdb:
```gdb
Thread 1 "python" received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50      ../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff7a81859 in __GI_abort () at abort.c:79
#2  0x00007ffff3a3d951 in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00007ffff3a4947c in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007ffff3a494e7 in std::terminate() () from /lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007ffff3a49799 in __cxa_throw () from /lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00007ffff5c0265f in cl::sycl::device_selector::select_device() const () from /opt/intel/oneapi/compiler/2021.1.1/linux/lib/libsycl.so.5
#7  0x00007ffff5c8581e in (anonymous namespace)::QueueManager::getQueueStack() () from /home/runner/work/dpctl/dpctl/dpctl/libDPCTLSyclInterface.so
#8  0x00007ffff5c85d28 in DPCTLQueueMgr_GetCurrentQueue () from /home/runner/work/dpctl/dpctl/dpctl/libDPCTLSyclInterface.so
#9  0x00007ffff2f72110 in __pyx_f_5dpctl_19_sycl_queue_manager_17_SyclQueueManager_get_current_queue (__pyx_v_self=<optimized out>, __pyx_skip_dispatch=__pyx_skip_dispatch@entry=1) at dpctl/_sycl_queue_manager.cpp:3206
#10 0x00007ffff2f72492 in __pyx_pf_5dpctl_19_sycl_queue_manager_17_SyclQueueManager_8get_current_queue (__pyx_v_self=<optimized out>) at dpctl/_sycl_queue_manager.cpp:3257
#11 __pyx_pw_5dpctl_19_sycl_queue_manager_17_SyclQueueManager_9get_current_queue (__pyx_v_self=<optimized out>, unused=<optimized out>) at dpctl/_sycl_queue_manager.cpp:3241
#12 0x00007ffff2f725c1 in __Pyx_PyObject_CallMethO (arg=0x0, func=0x7fffeb680f50) at dpctl/_sycl_queue_manager.cpp:20470
#13 __Pyx_PyObject_CallNoArg (func=0x7fffeb680f50) at dpctl/_sycl_queue_manager.cpp:20576
#14 __pyx_f_5dpctl_19_sycl_queue_manager_get_current_queue (__pyx_skip_dispatch=<optimized out>) at dpctl/_sycl_queue_manager.cpp:3893
#15 0x00007ffff2fc0108 in __pyx_f_5dpctl_6memory_7_memory_7_Memory__cinit_alloc (__pyx_v_self=0x7fffeb0a4d70, __pyx_v_alignment=0, __pyx_v_nbytes=8192, __pyx_v_ptr_type=0x7ffff5ca2d20, __pyx_v_queue=0x7ffff7f59710 <_Py_NoneStruct>)
    at dpctl/memory/_memory.cpp:4530
#16 0x00007ffff2fb9505 in __pyx_pf_5dpctl_6memory_7_memory_15MemoryUSMShared___cinit__ (__pyx_v_copy=<optimized out>, __pyx_v_queue=<optimized out>, __pyx_v_alignment=0, __pyx_v_other=<optimized out>, __pyx_v_self=0x7fffeb0a4d70)
    at dpctl/memory/_memory.cpp:8262
#17 __pyx_pw_5dpctl_6memory_7_memory_15MemoryUSMShared_1__cinit__ (__pyx_kwds=<optimized out>, __pyx_args=<optimized out>, __pyx_v_self=0x7fffeb0a4d70) at dpctl/memory/_memory.cpp:8215
#18 __pyx_tp_new_5dpctl_6memory_7_memory_MemoryUSMShared (t=<optimized out>, a=<optimized out>, k=<optimized out>) at dpctl/memory/_memory.cpp:22933
#19 0x00007ffff7d91a8a in type_call (type=type@entry=0x7ffff2fd1e80 <__pyx_type_5dpctl_6memory_7_memory_MemoryUSMShared>, args=args@entry=0x7fffeb0ac190, kwds=kwds@entry=0x0) at Objects/typeobject.c:951
```

### PR DESCRIPTION
**Pull request by diptorupd**
Closes #345 
Closes #344

- Fixes #345 and #344 by handling exceptions.
- Fixes C API tests that were segfaulting due to same reason. Now we skip if a device is not found.
- Add negative test cases to CAPI and Pytest to make sure we do not crash on getting an invalid filter.

Supersedes #348

### CODE REVIEW

**Comment by diptorupd** [comment]
I added checks to skip tests when a device does not exist. However, if current device does not exist then that is flagged as a failed test case.

**Review by oleksandr-pavlyk — APPROVED** [review]

File: dpctl-capi/source/dpctl_sycl_platform_interface.cpp [hunk 1]
```diff
@@ -84,6 +84,9 @@ __dpctl_give DPCTLSyclPlatformRef DPCTLPlatform_CreateFromSelector(
             PRef = wrap(P);
         } catch (std::bad_alloc const &ba) {
             std::cerr << ba.what() << '\n';
+        } catch (runtime_error const &re) {
+            std::cerr << re.what() << '\n';
+            return nullptr;
```
Comment (PokhodenkoSA) [inline dpctl-capi/source/dpctl_sycl_platform_interface.cpp:89]: Please, do `catch(...)` finally. Any uncatched exception will break C API interface.

**Review by PokhodenkoSA — CHANGES_REQUESTED** [review]
C API should not throw exceptions at all. We should catch all exceptions.

File: dpctl-capi/source/dpctl_sycl_platform_interface.cpp (hunk 1 above)
Comment (diptorupd) [reply dpctl-capi/source/dpctl_sycl_platform_interface.cpp:89]: Let us do that as a follow up along with the overall work needed to improve error handling (#35).

**Pull request merged**